### PR TITLE
Provider cycleclock() for 32bit ARM targets

### LINKED ISSUE #47347: Warning compiling on 32bit ARM systems: `No cycleclock() definition for your platform`
This doesn't seem to be a serious problem (for now, anyway), and of course this platform is Tier 3.. But I thought it would be good to highlight. Perhaps this way someone who cares about ARM 32bit sees it and is motivated to work on a fix :-)

Here is an example observed on `armv6l-linux-gnueabihf-cxx11` via Yggdrasil
```
[14:11:19] In file included from /workspace/srcdir/julia/src/gc.h:18:0,
[14:11:19]                  from /workspace/srcdir/julia/src/gc-debug.c:3:
[14:11:19] /workspace/srcdir/julia/src/julia_internal.h: In function 'cycleclock':
[14:11:19] /workspace/srcdir/julia/src/julia_internal.h:226:6: warning: #warning No cycleclock() definition for your platform [-Wcpp]
[14:11:19]      #warning No cycleclock() definition for your platform
[14:11:19]       ^~~~~~~
```

### PR DESCRIPTION
**Pull request by fingolfin**
Based on https://github.com/google/benchmark/blob/main/src/cycleclock.h

Resolves #47347

### CODE REVIEW

File: src/julia_internal.h [hunk 1]
```diff
@@ -216,6 +219,26 @@ static inline uint64_t cycleclock(void) JL_NOTSAFEPOINT
...
+    if (pmcntenset & 0x80000000ul) {  // Is it counting?
+      asm volatile("mrc p15, 0, %0, c9, c13, 0" : "=r"(pmccntr));
+      // The counter is set up to count every 64th cycle
+      return (int64_t)(pmccntr) * 64;  // Should optimize to << 6
+    }
+  }
+#endif
+  struct timeval tv;
+  gettimeofday(&tv, NULL);
+  return (int64_t)(tv.tv_sec) * 1000000 + tv.tv_usec;
```
Comment (ararslan) [inline src/julia_internal.h:241]: Indentation here should be 4 spaces
```suggestion
    // V6 is the earliest arch that has a standard cyclecount
#if (__ARM_ARCH >= 6)
    uint32_t pmccntr;
    uint32_t pmuseren;
    uint32_t pmcntenset;
    // Read the user mode perf monitor counter access permissions.
    asm volatile("mrc p15, 0, %0, c9, c14, 0" : "=r"(pmuseren));
    if (pmuseren & 1) {  // Allows reading perfmon counters for user mode code.
        asm volatile("mrc p15, 0, %0, c9, c12, 1" : "=r"(pmcntenset));
        if (pmcntenset & 0x80000000ul) {  // Is it counting?
            asm volatile("mrc p15, 0, %0, c9, c13, 0" : "=r"(pmccntr));
            // The counter is set up to count every 64th cycle
            return (int64_t)(pmccntr) * 64;  // Should optimize to << 6
        }
    }
#endif
    struct timeval tv;
    gettimeofday(&tv, NULL);
    return (int64_t)(tv.tv_sec) * 1000000 + tv.tv_usec;
```
Also, is `__ARM_ARCH` guaranteed to exist on ARM with both gcc and clang? Would a `defined(__ARM_ARCH)` be worthwhile to add?

Comment (ararslan) [reply src/julia_internal.h:241]: Should probably also reference the source of the code in a comment

Comment (fingolfin) [reply src/julia_internal.h]: Indentation is fixed.

As far as I can tell, all the `cycleclock` is taken from the same location, and there already is a comment indicating this (pointing to https://github.com/google/benchmark/blob/v1.5.0/src/cycleclock.h). Well, that comment is only in the fallback `#else` case; perhaps it should be moved to the top of `cycleclock` resp. into a comment before it? 

As to `__ARM_ARCH`, this is used unguarded in multiple places already. In any case, if it is not defined, then `#if (__ARM_ARCH >= 6)` will evaluate to `#if 0`, which arguably is the desired behaviour?

Comment (giordano) [reply src/julia_internal.h]: > Also, is `__ARM_ARCH` guaranteed to exist on ARM with both gcc and clang?

It's defined in the [Arm C Language Extensions](https://developer.arm.com/documentation/101028/latest) specification, so one would hope compliant compilers would follow it.

Comment (fingolfin) [reply src/julia_internal.h]: Sorry, I indeed missed the final three lines (and based on other surrounding code, the `#if` shouldn't be indented?). Hopefully OK now

**Review by ararslan — APPROVED** [review]
I don't have a 32-bit ARM system handy to verify myself but if it resolves the issue for you and doesn't obviously cause any others, this looks good to me. It correctly adjusts the C++ `static_cast` to a regular C cast and otherwise matches the linked code.

**Pull request merged**